core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -361,26 +361,16 @@ void mergeWith(
 
     Set<RelNode> parentRels = new HashSet<>(parents);
     for (RelNode otherRel : otherSet.rels) {
-      if (planner.prunedNodes.contains(otherRel)) {
-        continue;
-      }
-
-      boolean pruned = false;
       if (parentRels.contains(otherRel)) {
```

Comment:
Can you also add a check `otherRel` is not `spool` and is not an enforcer before `parentRels.contains(otherRel)`? I think we can't prune them in that case, what do you think?

Reply:
Maybe we can also remove https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java#L127-L130, they look duplicate with this part.

Reply:
done